Refactor release process documentation (#349)
kyma-project/community: guidelines/releases-guidelines/02-kyma-release-process.md

* A Git tag
* A release branch

## Preparation

> **NOTE:** This section applies only to new major and minor versions. If you release a patch, skip the preparation and go to the [**Steps**](#kyma-release-process-kyma-release-process-steps) section.

To prepare a release:

1. Create a release branch in the `kyma` repository. The name of this branch should follow the `release-x.y` pattern, such as `release-1.4`.

   ```bash
    git fetch upstream
    git checkout -b release-{release_version} upstream/master
    git push upstream release-{release_version}
    ```

   > **NOTE:** If you don't create the Kyma release branch at this point and add a  `post-rel{release_version}-kyma-release-candidate` post-submit job to the `test-infra` master, then pushing anything to the Kyma release branch, creating or rebasing the branch, triggers a new GitHub release.

2. [Define new release jobs](#kyma-release-process-kyma-release-process-preparation-define-new-release-jobs) in the `test-infra` repository.
3. [Remove old release jobs](#kyma-release-process-kyma-release-process-preparation-remove-old-release-jobs) in the `test-infra`.

### Define new release jobs

> Define release jobs on the `master` branch in the `test-infra` repository. To ensure every job name is unique, prefix it with `pre-rel{versionNumber}`. Remember to provide the version number without any periods. For example, to find all jobs for the 0.9 release, look for job names with the `pre-rel09` prefix. To learn how to define a release job for a component, read the following [document](https://github.com/kyma-project/test-infra/blob/master/docs/prow/release-jobs.md).

Reviewer: Should we change:
0.9 > `1.4`
pre-rel09 > `pre-rel14` ? 
since we were changing everything to `1.4` release version.

> 1. Navigate to the `test-infra` repository.

Reviewer: ```suggestion
1. Go to the `test-infra` repository.
```

2. Define release jobs in the `prow/jobs/test-infra` directory in the `watch-pods.yaml` file.
3. Define release jobs in the `prow/jobs/kyma` directory in the following files:

 * every `.yaml` in `components`
 * every `.yaml` in `tests`
 * every `.yaml` in `tools/` **except for** the `tools/failery/failery.yaml` file
 * `kyma-integration.yaml`
 * `kyma-artifacts.yaml`
 * `kyma-installer.yaml`
 * `kyma-release-candidate.yaml`
 * `kyma-github-release.yaml`

   > **NOTE:** Remember to modify the `presets` array. This is an example of the 1.4 release `preset`:

   ```yaml
   - labels:
       preset-target-commit-1.4: "true"
       env:
       - name: RELEASE_TARGET_COMMIT 
           value: release-1.4
   ```

>    > **NOTE:** Remember to update the `preset-target-commit-1.4` label for the `post-rel14-kyma-github-release` job. Follow the same procedure for each released version. See this example release 1.4.

Reviewer: ```suggestion
   > **NOTE:** Remember to update the `preset-target-commit-1.4` label for the `post-rel14-kyma-github-release` job. Follow the same procedure for each released version. For example, see release 1.4 configuration:
```

5. Update the `RELEASE_VERSION` file with the name of the next minor release candidate and merge the pull request to `master`. For example, if the `RELEASE_VERSION` on the `master` branch is set to `1.4.2`, then change the version to `1.5.0-rc1`.

### kyma-project/kyma

1. Inside the release branch do the following changes.

   i. Update your PR with the version and the directory of components used in `values.yaml` files.

   Find these values in the files:

   ```yaml
   dir: develop/
   version: {current_version}
   ```
  
   ```yaml
   dir: pr/
   version: {current_version}
   ```

   Replace them with:

   ```yaml
   dir:
   version: {release_version}
   ```

   > **NOTE:** Replace only `develop/` so `develop/tests` becomes `tests/`.

   Every component image is published with a version defined in the `RELEASE_VERSION` file stored in the `test-infra` repository on the given release branch. Test scripts for integration jobs like GKE Integration or GKE Upgrade are also loaded from the `test-infra` release branch.

>    For example, for the first release candidate of 1.1.0, the release version will be `1.1.0-rc1` and the `yaml` files should be modified as follows:

Reviewer: ```suggestion
   For example, for the first release candidate of 1.4.0, the release version will be `1.4.0-rc1` and the `yaml` files should be modified as follows:
```

   ``` yaml
   dir:

>    version: 1.1.0-rc1

Reviewer: ```suggestion
   version: 1.4.0-rc1
```

   ```

   > **CAUTION**: Do **not** update the version of components whose `dir` section does not contain `develop`, as is the case with Console-related components. Also do not change octopus version in `kyma/resources/testing/values.yaml` even though the directory is `develop` .

   ii. Check all `yaml` files in the `kyma` repository for references of the following Docker image:

   ```yaml
   image: eu.gcr.io/kyma-project/develop/{IMAGE_NAME}:{SOME_SHA}
   ```

   Change the Docker image to:

   ```yaml
   image: eu.gcr.io/kyma-project/{IMAGE_NAME}:{release_version}
   ```

   > **CAUTION**: In `installation/resources/installer.yaml` replace `eu.gcr.io/kyma-project/develop/installer:{image_tag}` with `eu.gcr.io/kyma-project/kyma-installer:{release_version}`

>    iii. In the `resources/core/values.yaml` file, replace the the `clusterDocsTopicsVersion` value with your release branch name. For example, for the 0.9.1 release, find the following section:

Reviewer: ```suggestion
   iii. In the `resources/core/values.yaml` file, replace the `clusterDocsTopicsVersion` value with your release branch name. For example, for the 1.4 release, find the following section:
```

   iv. Run the following tests in parallel:

     ```bash
     /test pre-rel14-kyma-gke-integration
     /test pre-rel14-kyma-gke-upgrade
     /test pre-rel14-kyma-gke-backup
     ```

   iv. Wait for the jobs to finish:
   * `pre-{release_number}-kyma-integration`
   * `pre-{release_number}-kyma-gke-integration`
   * `pre-{release_number}-kyma-gke-upgrade`
   * `pre-{release_number}-kyma-gke-backup`

5. If you detect any problems with the release, such as failing tests, wait for the fix that can be delivered either on a PR or cherry-picked to the PR from the `master` branch. Prow triggers the jobs again. Rerun manual jobs as described in **step 4**.

6. After all checks pass, merge the PR, using the `rebase and merge` option. To merge the PR to the release branch, you must receive approvals from all teams.

   > **CAUTION:** By default, the `rebase and merge` option is disabled. Contact one of the `kyma-project/kyma` repository admins to enable it.

7. Merging the PR to the release branch runs the postsubmit jobs, which:

   * create a GitHub release and trigger documentation update on the official Kyma website

>    * trigger provisioning of the cluster from the created release. The cluster name contains the release version with a period `.` replaced by a dash `-`. For example: `gke-release-1-1-0-rc1`. Use the cluster to test the release candidate.

Reviewer: ```suggestion
   * trigger provisioning of the cluster from the created release. The cluster name contains the release version with a period `.` replaced by a dash `-`. For example: `gke-release-1-4-0-rc1`. Use the cluster to test the release candidate.
```
